mail/mailer_test.go

```diff
@@ -196,23 +204,69 @@ func badEmailHandler(messagesToProcess int) connHandler {
 	}
 }
 
-func setup(t *testing.T) (*MailerImpl, net.Listener, func()) {
+// The rstHandler authenticates the client like the normalHandler but
+// additionally processes an email flow (e.g. MAIL, RCPT and DATA
+// commands). When the `connID` is <= `rstFirst` the socket of the
+// listening connection is set to abruptively close (sends TCP RST but
+// no FIN). The listening connection is closed immediately after the
+// MAIL command is received and prior to issuing a 250 response. In this
+// way the first `rstFirst` connections will not complete normally and
+// can be tested for reconnection logic.
+func rstHandler(rstFirst int) connHandler {
+	return func(connID int, t *testing.T, tlsConn net.Conn, tcpConn *net.TCPConn) {
+		defer func() {
+			err := tcpConn.Close()
+			if err != nil {
+				t.Errorf("conn.Close: %s", err)
+			}
+		}()
+		authenticateClient(t, tlsConn)
+
+		buf := bufio.NewReader(tlsConn)
+		if err := expect(t, buf, "MAIL FROM:<<you-are-a-winner@example.com>> BODY=8BITMIME"); err != nil {
+			return
+		}
+		// Set the socket of the listening connection to abruptively
+		// close.
+		if connID <= rstFirst {
+			err := tcpConn.SetLinger(0)
+			if err != nil {
+				return
+			}
+			fmt.Printf("Socket set for abruptive close. Cutting off client early\n")
```

Comment:
In tests we should use `t.Log` (and note that `\n` isn't needed in t.Log)

Reply:
I was matching to local style, I'll go ahead and fix the other lines as well.